packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -405,75 +434,71 @@ export class Pipeline extends PipelineBase {
-      return {
-        artifactBucket: this.artifactBucket,
-      };
-    }
-    // get the region the Pipeline itself is in
-    const pipelineRegion = this.requireRegion();
-    // if the action is in the same region as the pipeline, nothing to do
-    if (actionRegion === pipelineRegion) {
+    if (anyRegion || sameEnvDimension(this.env.region, region)) {
       return {
         artifactBucket: this.artifactBucket,
       };
     }
 
+    // The action has a region and it's not the pipeline's, require the pipeline
+    // to have a known region as well.
+    this.requireRegion();
+
     // source actions have to be in the same region as the pipeline
     if (action.actionProperties.category === ActionCategory.SOURCE) {
       throw new Error(`Source action '${action.actionProperties.actionName}' must be in the same region as the pipeline`);
     }
 
     // check whether we already have a bucket in that region,
     // either passed from the outside or previously created
+    const crossRegionSupport = this.obtainCrossRegionSupportFor(action, region);
+
+    // the stack containing the replication bucket must be deployed before the pipeline
+    Stack.of(this).addDependency(crossRegionSupport.stack);
+    // The Pipeline role must be able to replicate to that bucket
+    crossRegionSupport.replicationBucket.grantReadWrite(this.role);
+
+    return {
+      artifactBucket: crossRegionSupport.replicationBucket,
+      region,
+    };
+  }
+
+  /**
+   * Get or create the cross-region support construct for the given action
+   */
+  private obtainCrossRegionSupportFor(action: IAction, actionRegion: string) {
     let crossRegionSupport = this._crossRegionSupport[actionRegion];
     if (!crossRegionSupport) {
       // we need to create scaffolding resources for this region
+      const otherStack = this.scopeForCrossRegionSupportConstruct(action);
       crossRegionSupport = this.createSupportResourcesForRegion(otherStack, actionRegion);
       this._crossRegionSupport[actionRegion] = crossRegionSupport;
     }
+    return crossRegionSupport;
+  }
 
-    // the stack containing the replication bucket must be deployed before the pipeline
-    pipelineStack.addDependency(crossRegionSupport.stack);
-    crossRegionSupport.replicationBucket.grantReadWrite(this.role);
+  /**
+   * Return the correct scope to create the cross-region construct in
+   *
+   * As an optimization, we reuse the action resource's stack if the stack has
+   * the same environment as the action.
+   *
+   * (Always true for owned action resources but not necessarily true
+   * for imported ones).
+   */
+  private scopeForCrossRegionSupportConstruct(action: IAction): Stack | undefined {
+    const actionResource = action.actionProperties.resource;
+    if (!actionResource) { return undefined; }
 
-    return {
-      artifactBucket: crossRegionSupport.replicationBucket,
-      region: actionRegion,
-    };
+    const stack = Stack.of(actionResource);
```

Comment:
Can we call these `actionResourceStack` and `actionResourceStackEnv`, respectively? Then, the code becomes

```ts
    return sameEnv(actionResource.env, actionResourceStackEnv) ? actionResourceStack : undefined
```

which reads perfectly to me.